3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -351,3 +479,69 @@ case class PiiStrategyPseudonymize(hashFunction: MessageDigest) extends PiiStrat
   override def scramble(clearText: String): String = hash(clearText)
   def hash(text: String): String                   = String.format("%064x", new java.math.BigInteger(1, hashFunction.digest(text.getBytes(TextEncoding))))
 }
+
+/**
+ * The mutator class encapsulates the mutator function and the field name where the mutator will be applied.
+ */
+case class Mutator(fieldName: String, muatatorFn: PiiConstants.MutatorFn)
+
+/**
+ * Case class for capturing scalar field modifications.
+ */
+case class ScalarModifiedField(fieldName: String, originalValue: String, modifiedValue: String) extends ModifedField
+
+/**
+ * Case class for capturing JSON field modifications.
+ */
+case class JsonModifiedField(field: String, originalValue: String, modifiedValue: String, jsonPath: String, schema: String)
+    extends ModifedField
+
+/**
+ * Parent class for classes that serialize the values that were modified during the PII enrichment.
+ */
+case class PiiModifiedFields(val modifiedFields: PiiConstants.ModifiedFields, val strategy: PiiStrategy)
+
+/**
+ * Custom serializer for PiiStrategy class
+ */
+class PiiStrategySerializer
+    extends CustomSerializer[PiiStrategy](format =>
+      ({
+        case jo: JObject =>
+          implicit val json4sFormats = DefaultFormats
+          val function               = (jo \ "pseudonymize" \ "hashFunction").extract[String]
+          PiiStrategyPseudonymize(MessageDigest.getInstance(function))
+      }, {
+        case psp: PiiStrategyPseudonymize =>
+          "pseudonymize" -> ("hashFunction" -> psp.hashFunction.getAlgorithm)
+      }))
+
+/**
+ * Custom serializer for PiiModifiedFields class
+ */
+class PiiModifiedFieldsSerializer
+    extends CustomSerializer[PiiModifiedFields](format => {
+      val PiiTransformationSchema = "iglu:com.snowplowanalytics.snowplow/pii_transformation/jsonschema/1-0-0"
+      ({
+        case jo: JObject =>
+          implicit val json4sFormats = DefaultFormats + new PiiStrategySerializer
+          val fields                 = (jo \ "data" \ "pii").extract[List[ModifedField]]
+          val strategy               = (jo \ "data" \ "strategy").extract[PiiStrategy]
+          PiiModifiedFields(fields, strategy)
+      }, {
+        case pmf: PiiModifiedFields =>
+          implicit val json4sFormats = DefaultFormats + new PiiStrategySerializer
+          ("schema" -> PiiTransformationSchema) ~
+            ("data" ->
+              ("pii" -> decompose(
+                pmf.modifiedFields
+                  .map {
+                    case s: ScalarModifiedField => "pojo" -> s
+                    case j: JsonModifiedField   => "json" -> j
+                  }
+                  .groupBy(_._1)
+                  .mapValues(_.map(_._2))
```

Comment:
can't we extract this data wrangling from the json construction? this will make thing easier to understand

Reply:
also can we turn this into a fold, that'll be more performant

Reply:
As far as I can tell from bench marking this is not more performant. I tried your example and it was actually slower.

Reply:
Have you read:

>I reliably get those results https://gist.github.com/BenFradet/13c4a33ea6c0a850de8728f4cca3d82c
>anyway doing one pass instead of two over the data is preferable imo
>plus you avoid creating useless stuff

Reply:
Anyways ... you can even save another pass by doing:

```scala
pmf.modifiedFields.foldLeft(Map.empty[String, List[ModifiedField]]) { case (m, mf) =>
  mf match {
    case s: ScalarModifiedField => m + ("pojo" -> (s :: m.getOrElse("pojo", List.empty[MF])))
    case j: JsonModifiedField => m + ("json" -> (j :: m.getOrElse("json", List.empty[MF])))
  }
}
```

Reply:
This comment was lost here due to the update. I just found it in my emails